### DISCUSSION TITLE
DAI 2023-2024 - Java IOs benchmarking - Alexandre Philibert

### DISCUSSION THREAD

**Discussion by AlexandrePhilibert** [heig-vd-dai-course/heig-vd-dai-course#72] [Show and tell]
[Repo](https://github.com/AlexandrePhilibert/dai-course-java-ios)

###  Which type of stream is the most efficient for each use case?

Buffered

### Why is it more efficient than the other types of streams?

Because we do not do a syscall for each byte, thus reducing the overhead of each `.write()` or `.read()` calls

### What is the difference between binary data and text data?

One is human readable

### What is a character encoding?

A character encoding specifies how to "translate" bytes into text and the other way.

### Why is this methodology important?

Understanding how to best use the tools at our disposal is important

**Text Data**

| Size     | Buffered Write   | Unbuffered Write | Buffered Read  | Unbuffered Read |
|----------|-------------------|-------------------|----------------|-----------------|
| 1B       | 191990ns          | 374988ns          | 323098ns       | 787075ns        |
| 1024B    | 678495ns          | 3211851ns         | 705616ns       | 2794606ns       |
| 1MB      | 14ms              | 66ms              | 21ms           | 50ms            |
| 5MB      | 103ms             | 175ms             | 90ms           | 152ms           |

**Binary Data**

| Size     | Buffered Write   | Unbuffered Write | Buffered Read  | Unbuffered Read |
|----------|-------------------|-------------------|----------------|-----------------|
| 1B       | 411293ns          | 299945ns          | 134283ns       | 156695ns        |
| 1024B    | 105153ns          | 98899ns           | 426655ns       | 1368098ns       |
| 1MB      | 1ms               | 2ms               | 20ms           | 943ms           |
| 5MB      | 5ms               | 9ms               | 96ms           | 4628ms          |

© ChatGPT for the tables

**Comment by ludelafo** [comment]
Thank you for the mention of the tables' source! Great usage of ChatGPT.

I would add that this scientific methodology to benchmark our applications can ensure the performance are still good over time for the use cases we are managing.

**EDIT**: For text files, do not forget to specify the charset of the file (while reading and writing!). It is more explicit and you can avoid many issues in the future.

One of the constructors for the `FileReader` et `FileWriter` classes allows to specify the charset to ensure the characters are interpreted correctly by your application: 

- https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/FileReader.html
- https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/FileWriter.html
